usdQt/undoStateDelegate.cpp

```diff
@@ -0,0 +1,405 @@
+        _RouteInverse(std::bind(&UsdQt_UndoStateDelegate::_InvertSetField,
+                                this, id.GetFullSpecPath(),
+                                SdfFieldKeys->TimeSamples, VtValue()));
+    } else {
+        VtValue oldValue;
+        _GetLayer()->QueryTimeSample(id, time, &oldValue);
+        _RouteInverse(
+            std::bind(&UsdQt_UndoStateDelegate::_InvertSetTimeSample, this,
+                      id.GetFullSpecPath(), time, oldValue));
+    }
+}
+
+void UsdQt_UndoStateDelegate::_OnSetField(const SdfAbstractDataSpecId& id,
+                                               const TfToken& fieldName,
+                                               const VtValue& value) {
+    _OnSetFieldImpl(id, fieldName);
+}
+
+void UsdQt_UndoStateDelegate::_OnSetField(
+    const SdfAbstractDataSpecId& id, const TfToken& fieldName,
+    const SdfAbstractDataConstValue& value) {
+    _OnSetFieldImpl(id, fieldName);
+}
+
+void UsdQt_UndoStateDelegate::_OnSetFieldDictValueByKey(
+    const SdfAbstractDataSpecId& id, const TfToken& fieldName,
+    const TfToken& keyPath, const VtValue& value) {
+    _OnSetFieldDictValueByKeyImpl(id, fieldName, keyPath);
+}
+
+void UsdQt_UndoStateDelegate::_OnSetFieldDictValueByKey(
+    const SdfAbstractDataSpecId& id, const TfToken& fieldName,
+    const TfToken& keyPath, const SdfAbstractDataConstValue& value) {
+    _OnSetFieldDictValueByKeyImpl(id, fieldName, keyPath);
+}
+
+void UsdQt_UndoStateDelegate::_OnSetTimeSample(
+    const SdfAbstractDataSpecId& id, double time, const VtValue& value) {
+    _OnSetTimeSampleImpl(id, time);
+}
+
+void UsdQt_UndoStateDelegate::_OnSetTimeSample(
+    const SdfAbstractDataSpecId& id, double time,
+    const SdfAbstractDataConstValue& value) {
+    _OnSetTimeSampleImpl(id, time);
+}
+
+void UsdQt_UndoStateDelegate::_OnCreateSpec(const SdfPath& path,
+                                                 SdfSpecType specType,
+                                                 bool inert) {
+    TF_DEBUG(USDQT_DEBUG_UNDOSTATEDELEGATE)
+        .Msg("Creating spec at '%s'\n", path.GetText());
+    _MarkCurrentStateAsDirty();
+
+    _RouteInverse(std::bind(&UsdQt_UndoStateDelegate::_InvertCreateSpec,
+                            this, path, inert));
+}
+
+// XXX: This is copied from SdLayer
+static void _CopySpecAtPath(const SdfAbstractData& src, SdfAbstractData* dst,
+                            const SdfPath& path) {
+    _CopySpec(src, dst, SdfAbstractDataSpecId(&path));
+}
+
+void UsdQt_UndoStateDelegate::_OnDeleteSpec(const SdfPath& path,
+                                                 bool inert) {
+    TF_DEBUG(USDQT_DEBUG_UNDOSTATEDELEGATE)
+        .Msg("Deleting spec at '%s'\n", path.GetText());
+    _MarkCurrentStateAsDirty();
+
+    SdfDataRefPtr deletedData = TfCreateRefPtr(new SdfData());
+    SdfLayer::TraversalFunction copyFunc = boost::bind(
+        &::_CopySpecAtPath, boost::cref(*boost::get_pointer(_GetLayerData())),
```

Comment:
`::_CopySpecAtPath` -> `_CopySpecAtPath`

Reply:
Why are you advocating for not using ::?

Reply:
Sorry, I should have been more clear.  When using the pxr namespace `_CopySpecAtPath` cannot be found in `::`.